### PR TITLE
Fix error parsing subtree when loading from file path

### PR DESCRIPTION
**Pull request by mini-1235**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | Fixes https://github.com/ros-navigation/navigation2/issues/5923 |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | (Steve's Robot, gazebo simulation of Tally, hardware turtlebot) |
| Does this PR contain AI generated software? | (No; Yes and it is marked inline in the code) |
| Was this PR description generated by AI software? | Out of respect for maintainers, AI for human-to-human communications are banned |

---

## Description of how this change was tested

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists
- [ ] Should this be backported to current distributions? If so, tag with `backport-*`.

### CODE REVIEW

**Comment by mini-1235** [comment]
This PR reverts #5938 and #5594.

I believe the correct logic should be:
- When file path is provided, we should call `createTreeFromFile` directly, which matches the behavior prior to https://github.com/ros-navigation/navigation2/pull/5426
- When ID is provided, we loop through the `search_directory`, register the trees, and then create the tree using the specified ID

**Comment by adivardi** [comment]
Thanks @mini-1235 

Tested:
- With all BTs having unique IDs
	- requesting both filepath & ID work, with subTree both before and after the main tree
- With some duplicated IDs
  - requesting filepath work
  - requesting ID: Actually also work, but I am getting the confusing warning again `Duplicate BT IDs found. Make sure all
 BT IDs are unique! ID: MainTree File:...`
		- maybe that's intended? I don't have multiple BTs for the same action to test though, only multiple ones for different navigators.

**Comment by mini-1235** [comment]
> requesting ID: Actually also work, but I am getting the confusing warning again Duplicate BT IDs found. Make sure all  BT IDs are unique! ID: MainTree File:...
> - maybe that's intended? I don't have multiple BTs for the same action to test though, only multiple ones for different navigators.

Yeah, I think this is expected since it can be confusing which BT is being referenced if there are duplicate ids and the tree is requested by id. In that case, I would expect the user to see this warning and update their BTs accordingly. But I am open to suggestions on how to make this clearer, perhaps doing something similar to what was proposed in #5938?

**Comment by mini-1235** [comment]
The other failing test seems to be related to https://github.com/ros-navigation/navigation2/commit/797a1719231babee67f3b1d62d6457032bc4b21a

**Comment by adivardi** [comment]
> Yeah, I think this is expected since it can be confusing which BT is being referenced if there are duplicate ids and the tree is requested by id. In that case, I would expect the user to see this warning and update their BTs accordingly. But I am open to suggestions on how to make this clearer, perhaps doing something similar to what was proposed in #5938?

I think that would be nice. Thought it is already better since it doesn't have the `skipping BT file` while not actually skipping anything :hand_over_mouth:

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp [hunk 1]
```diff
@@ -236,10 +236,50 @@ void BtActionServer<ActionT, NodeT>::setGrootMonitoring(
...
+              logger_,
+              "Skipping BT file %s (no valid BehaviorTree IDs found)",
+              entry.path().string().c_str());
+            continue;
+          }
+          // Check for duplicate BT IDs across files and warn if any are found
+          for (const auto & current_bt_id : bt_ids) {
+            auto [it, inserted] = used_bt_id.insert(current_bt_id);
+            if (!inserted) {
+              RCLCPP_WARN(
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp]: I consolidated all the logging here to a single log from #5938 : https://github.com/ros-navigation/navigation2/blob/main/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp#L322-L336 that would be nice as well so its not logged N times.

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp [hunk 2]
```diff
@@ -260,91 +300,22 @@ bool BtActionServer<ActionT, NodeT>::loadBehaviorTree(const std::string & bt_xml
...
-      // When a filename is specified, it must be register first
-      // and treat it as the "main" tree to execute.
-      // This ensures the requested tree is always available
-      // and prioritized, even if other files in the directory have duplicate IDs.
-      // The lambda then skips this main file to avoid
-      // re-registering it or logging a duplicate warning.
-      // In contrast, when an ID is specified, it's unknown which file is "main"
-      // so all files are registered and conflicts are handled in the lambda.
-      register_all_bt_files(main_file);
+      tree_ = bt_->createTreeFromFile(file_or_id, blackboard_);
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp]: Doesn't this not support subtrees in other files within a file-based call?

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp [hunk 3]
```diff
@@ -260,91 +300,22 @@ bool BtActionServer<ActionT, NodeT>::loadBehaviorTree(const std::string & bt_xml
...
-      // file_or_id is a filename: register it first
-      std::string main_file = file_or_id;
-      main_id = bt_->extractBehaviorTreeID(main_file);
-      if (main_id.empty()) {
-        RCLCPP_ERROR(logger_, "Failed to extract ID from %s", main_file.c_str());
+    if(!is_bt_id) {
+      // Read the input BT XML from the specified file into a string
+      std::ifstream xml_file(file_or_id);
+      if (!xml_file.good()) {
+        RCLCPP_ERROR(logger_, "Couldn't open input XML file: %s", file_or_id.c_str());
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp]: Doesn't the catch cover this? Also if it fails, wouldn't this be a `setInternalError` worthy thing? So we should let the exception get thrown

Comment (mini-1235) [reply nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp]: will update this

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp (hunk 1 above)
Comment (mini-1235) [reply nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp]: will update this

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp (hunk 2 above)
Comment (mini-1235) [reply nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp]: When using file path, I would expect users to `<include>` in their BT XML files, as described here: https://www.behaviortree.dev/docs/tutorial-basics/tutorial_07_multiple_xml#add-multiple-files-with-include, this matches the behavior we had before ID based loading was introduced

I also don’t have a good solution in mind how we could register all behavior trees and then call `createTree` when loading from a file path

Comment (SteveMacenski) [reply nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp]: I'm not saying this is right or wrong, but before we had:

```

      // When a filename is specified, it must be register first
      // and treat it as the "main" tree to execute.
      // This ensures the requested tree is always available
      // and prioritized, even if other files in the directory have duplicate IDs.
      // The lambda then skips this main file to avoid
      // re-registering it or logging a duplicate warning.
      // In contrast, when an ID is specified, it's unknown which file is "main"
      // so all files are registered and conflicts are handled in the lambda.
      register_all_bt_files(main_file);
```

And the `register_all_bt_files` lambda that you refactored into a function took that `main_file` as a skip key, so we'd still register all the files *except* `main_file` in the case of loading by filepath. That way subtrees in other files were registered for use.

Comment (mini-1235) [reply nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp]:
```
 // When a filename is specified, it must be register first
      // and treat it as the "main" tree to execute.
 ```
 
This seems to conflict with what this PR is trying to address. If a file contains multiple <BehaviorTree> tags, how do we determine which one should be used when calling:

```
try {
    tree_ = bt_->createTree(main_id, blackboard_);
```

I don't have an easy way to do this, any suggestions?

Comment (SteveMacenski) [reply nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp]: Why not something like:

* Extract the IDs (plural) from the main file
* register those
* Call `registerAllTreesFromDirectories` skipping the IDs extracted from the main file
* Then finally `createTree`

That mirrors what is there now except registering all trees in the main file and skipping all trees in teh main file in otherwise registration to handle conflicts correctly. 

> how do we determine which one should be used when calling

How does it know now?

Comment (mini-1235) [reply nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp]: > Why not something like:

I think that is similar to what we had before. The problem is that it treats the first `<BehaviorTree`> tag as "main_id", as shown in the log below:

```
[bt_navigator-23] [INFO 1770808676.977817977] [bt_navigator]: Created BT from ID: Subtree (loadBehaviorTree() at /root/nav2_ws/src/navigation2/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp:347)
```

https://github.com/ros-navigation/navigation2/blob/064ca7585523f5205328eb186f43b65a52de428d/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp#L346-L360

> How does it know now?

`createTreeFromFile` supports multiple `<BehaviorTree>` tags correctly when `main_tree_to_execute` is specified, it uses that attribute to determine which tree should be instantiated. So I think one possible approach to find "main_id" would be:


- Extract `main_tree_to_execute` first (optional tag)
- If it doesn't exist, then fallback to the first `<BehaviorTree>` tag 


How does that sound to you?

Comment (mini-1235) [reply nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp]: I have spent some time working on another prototype, please take a look

ps: I haven't tested it yet, I will test and add code coverage later if you agree with the general idea

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp [hunk 4]
```diff
@@ -273,19 +273,29 @@ bool BtActionServer<ActionT, NodeT>::loadBehaviorTree(const std::string & bt_xml
             continue;
           }
 
-          auto id = bt_->extractBehaviorTreeID(entry.path().string());
-          if (id.empty()) {
+          auto tree_info = bt_->parseTreeInfo(entry.path().string());
+          if (tree_info.all_ids.empty()) {
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp]: main_id instead?

File: nav2_behavior_tree/src/behavior_tree_engine.cpp [hunk 5]
```diff
@@ -105,46 +105,44 @@ BehaviorTreeEngine::createTreeFromFile(
...
-    RCLCPP_ERROR(
-      rclcpp::get_logger("BehaviorTreeEngine"),
-      "Error: <BehaviorTree> element not found in %s", bt_file.c_str());
-    return "";
+
+  tinyxml2::XMLElement * root = doc.RootElement();
+  if (!root) {return info;}
+
+  // First try to get main_tree_to_execute attribute
+  const char * main_attr = root->Attribute("main_tree_to_execute");
```
Comment (SteveMacenski) [inline nav2_behavior_tree/src/behavior_tree_engine.cpp]: From BT.CPP:

> The attribute [main_tree_to_execute] is mandatory if the file contains multiple <BehaviorTree>, optional otherwise.

We can hard enforce this constraint when multiple exist and fail the load with an error.

File: nav2_behavior_tree/src/behavior_tree_engine.cpp [hunk 6]
```diff
@@ -105,46 +105,44 @@ BehaviorTreeEngine::createTreeFromFile(
...
-    return "";
+
+  // Loop through all BehaviorTree elements to get all IDs
+  for (auto * bt = root->FirstChildElement("BehaviorTree"); bt;
+    bt = bt->NextSiblingElement("BehaviorTree"))
+  {
+    const char * id = bt->Attribute("ID");
+    if (id) {
+      info.all_ids.emplace_back(id);
+      // Fallback if no main_tree_to_execute was set, the first BT found becomes the main_id
```
Comment (SteveMacenski) [inline nav2_behavior_tree/src/behavior_tree_engine.cpp]: So that we don't have to have the opinionated policy; it must be set if multiple trees are in the file

File: nav2_behavior_tree/src/behavior_tree_engine.cpp [hunk 7]
```diff
@@ -105,46 +105,44 @@ BehaviorTreeEngine::createTreeFromFile(
...
-  }
-  tinyxml2::XMLElement * btElement = rootElement->FirstChildElement("BehaviorTree");
-  if (!btElement) {
-    RCLCPP_ERROR(
-      rclcpp::get_logger("BehaviorTreeEngine"),
-      "Error: <BehaviorTree> element not found in %s", bt_file.c_str());
-    return "";
+
+  tinyxml2::XMLElement * root = doc.RootElement();
+  if (!root) {return info;}
```
Comment (SteveMacenski) [inline nav2_behavior_tree/src/behavior_tree_engine.cpp]: Log please

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp (hunk 4 above)
Comment (mini-1235) [reply nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp]: ~~What about renaming `all_ids` to `behavior_tree_ids`, so in the new struct BTInfo:~~

~~main_id -> the ID provided in `main_tree_to_execute`~~
~~behavior_tree_ids -> all `<BehaviorTree>` tag IDs~~

Can you clarify? I don't think we are using main_id in this function

File: nav2_behavior_tree/src/behavior_tree_engine.cpp (hunk 5 above)
Comment (mini-1235) [reply nav2_behavior_tree/src/behavior_tree_engine.cpp]: Agree, check latest commit

File: nav2_behavior_tree/src/behavior_tree_engine.cpp (hunk 6 above)
Comment (mini-1235) [reply nav2_behavior_tree/src/behavior_tree_engine.cpp]: Agree, check latest commit

File: nav2_behavior_tree/src/behavior_tree_engine.cpp (hunk 7 above)
Comment (mini-1235) [reply nav2_behavior_tree/src/behavior_tree_engine.cpp]: See latest commit

**Comment by mini-1235** [comment]
> Tested:
With all BTs having unique IDs
requesting both filepath & ID work, with subTree both before and after the main tree
With some duplicated IDs
requesting filepath work
requesting ID: Actually also work, but I am getting the confusing warning again Duplicate BT IDs found. Make sure all  BT IDs are unique! ID: MainTree File:...

I tested these again with my latest version, and everything looks fine to me

**Comment by adivardi** [comment]
> I tested these again with my latest version, and everything looks fine to me

I didn't do a full test again now, just tested our main xml file, and it is working with this PR, even when requesting a filename

**Comment by mini-1235** [comment]
> > I tested these again with my latest version, and everything looks fine to me
> 
> I didn't do a full test again now, just tested our main xml file, and it is working with this PR, even when requesting a filename

Thank you @adivardi !

File: nav2_behavior_tree/src/behavior_tree_engine.cpp [hunk 8]
```diff
@@ -105,46 +105,62 @@ BehaviorTreeEngine::createTreeFromFile(
...
+      // which one to use as the main tree
+      if (info.behavior_tree_ids.size() == 1) {
+        info.main_id = info.behavior_tree_ids[0];
+      } else if (info.behavior_tree_ids.size() > 1) {
+        RCLCPP_ERROR(
+          rclcpp::get_logger("BehaviorTreeEngine"),
+          "Multiple BehaviorTree elements found in %s "
+          "but no main_tree_to_execute attribute specified.",
+          filename.c_str());
+      }
```
Comment (SteveMacenski) [inline nav2_behavior_tree/src/behavior_tree_engine.cpp]: Should this throw/fail as an invalid BT file? The BT.CPP rules states it must be (and disambiguates for us)

File: nav2_behavior_tree/src/behavior_tree_engine.cpp [hunk 9]
```diff
@@ -105,46 +105,62 @@ BehaviorTreeEngine::createTreeFromFile(
...
-  if (idValue) {
-    return std::string(idValue);
-  } else {
-    RCLCPP_ERROR(
-      rclcpp::get_logger("BehaviorTreeEngine"),
-      "Error: ID attribute not found on <BehaviorTree> element in %s",
-      bt_file.c_str());
-    return "";
+
+  if (main_id_required) {
```
Comment (SteveMacenski) [inline nav2_behavior_tree/src/behavior_tree_engine.cpp]: Then I think this argument and check can be removed. We always populate with either `main_tree_to_execute`, the single-instance tree, or fail.

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp (hunk 4 above)
Comment (SteveMacenski) [reply nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp]: Shouldn't the `main_id` be checked as required, not `all_ids`? But I think from the recent comments below, this point is much less important - semantically it will have the same effect

File: nav2_behavior_tree/src/behavior_tree_engine.cpp (hunk 9 above)
Comment (mini-1235) [reply nav2_behavior_tree/src/behavior_tree_engine.cpp]: Also responding to https://github.com/ros-navigation/navigation2/pull/5950#discussion_r2817948806

My understanding is that `main_tree_to_execute` is only required when using file path and that file contains multiple `<BehaviorTree>` tags, then we need to specify which tree should be executed as the main one. However, if we are specifying the tree directly by ID, then `main_tree_to_execute` is not necessary, since the main_id is already known

So I am adding an argument here, I believe there are some cases where we don't actually care about which tree is the main one, instead we only care about what `<BehaviorTree>` tags are defined in the file (aka the IDs that are registered)

File: nav2_behavior_tree/src/behavior_tree_engine.cpp (hunk 8 above)
Comment (mini-1235) [reply nav2_behavior_tree/src/behavior_tree_engine.cpp]: But I agree that we should probably throw here

Comment (SteveMacenski) [reply nav2_behavior_tree/src/behavior_tree_engine.cpp]: Notably also in `nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp` 's `if (tree_info.main_id.empty()) {` we return false but don't set the internal error code.

File: nav2_behavior_tree/src/behavior_tree_engine.cpp (hunk 9 above)
Comment (SteveMacenski) [reply nav2_behavior_tree/src/behavior_tree_engine.cpp]: True, though Davide's rules say that a file with N trees needs to have one marked as `main_tree_to_execute`. Even if we know which we want to run, its still a malformed file. I like being a stickler for this kind of thing, it disambiguates problems before they arise and cause subtle issues. For instance, while it works for loading by ID, the same tree would fail loading by XML path, so that's unexpected outcome for users.

Comment (mini-1235) [reply nav2_behavior_tree/src/behavior_tree_engine.cpp]: That makes sense, I will update

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp (hunk 4 above)
Comment (mini-1235) [reply nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp]: We are looping through all ids to check if it has been registered, so I think we should still check all instead of main here

**Comment by mini-1235** [comment]
022477e added some test cases, hopefully this covers all the corner cases :)

**Review by SteveMacenski — APPROVED** [review]
@adivardi are you happy with the final changes as well?

**Comment by adivardi** [comment]
In general looks good, also tested a few scenarios and it is working fine.
The `Skipping conflicting BT XML files` shows up correctly when I have conflicting IDs.

**Comment by SteveMacenski** [comment]
Great, thanks for being persistent on this issue & thanks @mini-1235 for handling this!
